### PR TITLE
nimble: add monitor protocol/interface

### PR DESCRIPTION
**Pull request by andrzej-kaczmarek**
See here for details: http://mail-archives.apache.org/mod_mbox/incubator-mynewt-dev/201706.mbox/%3CCADAkoF4LQOdqH%3DLh%3D5GgBPC9nEaXV95TV_uQPXXWuhrpCXfbtQ%40mail.gmail.com%3E

This is core monitor protocol functionality extracted from https://github.com/apache/mynewt-core/pull/362 so logging stuff does not block it.

### CODE REVIEW

**Comment by rymanluk** [comment]
Changes look good

File: net/nimble/host/src/ble_monitor.c [hunk 1]
```diff
@@ -243,4 +257,58 @@ ble_monitor_new_index(uint8_t bus, uint8_t *addr, const char *name)
     return 0;
 }
 
+int
+ble_monitor_log(int level, const char *fmt, ...)
+{
+    struct ble_monitor_hdr hdr;
+    struct ble_monitor_user_logging ulog;
+    const char id[] = "nimble";
```
Comment (sjanc) [inline net/nimble/host/src/ble_monitor.c]: could be static const

**Comment by andrzej-kaczmarek** [comment]
@ccollins476ad 
regarding vsnprintf(): C99 guarantees that with size=0 nothing is written to buffer and we can pass buffer as NULL so I used it this way. This however required fixing of tinyprintf's implementation of vsnprintf() which I also included here.

**Comment by ccollins476ad** [comment]
Thanks, Andrzej.  It looks good.  Also, thanks for clarifying the snprintf() semantics.  I should have checked the spec.  Just for the record, from n1256:

```
7.19.6.5 The snprintf function

Synopsis

    #include <stdio.h>
    int snprintf(char * restrict s, size_t n,
         const char * restrict format, ...);

Description

The snprintf function is equivalent to fprintf, except that the output is written into
an array (specified by argument s) rather than to a stream. If n is zero, nothing is written,
and s may be a null pointer.
```

**Pull request merged**